python/triton/runtime/jit.py

```
@@ -566,6 +566,9 @@ def run(self, *args, grid, warmup, **kwargs):
                 return None
             # compile the kernel
             src = self.ASTSource(self, signature, constexprs, attrs)
+            # cap warp count to 16 for devices with warp size 64
+            if target.warp_size == 64:
+                options.__dict__.update({'num_warps': 16})
```

so this just overrides `num_warps` to 16? How can it be correct? cc: @antiagainst could you help figure out a cleaner way? I don't think we should change the num_warps asked for user. If the value is invalid we should error and ideally do it in the backend specific code.

---

Realized this should check if warp size == 64 and num_warps > 16, shouldn't override unless it's actually invalid. That being said I don't know the full impact overriding the user's request could have, if there's a cleaner/saner/more performant way to do this please suggest.

---

I still don't think silently changing num_warp picked by user is right

---

Yes we should error out instead of silently changing what's provided by the user.

---

Sure, will change to an assert; won't directly address the issue (which is really on the workload's side), but will at least provide proper guidance on how to fix it.